[rtsan] Enable RealtimeSanitizer for FreeBSD

devnexen opened the pull request.

devnexen committed:
[compiler-rt][rtsan] porting the sanitizer to FreeBSD.

Most of the apple api exceptions also apply to freebsd, however
to create a per-thread realtime context pthread api cannot be used
since freebsd calls calloc in _thr_alloc().

devnexen committed:
freebsd clang frontend update.

davidtrevelyan commented:
@devnexen in principle, I'm of course not opposed to rtsan supporting FreeBSD - is there any motivation in particular for why you propose adding it here? Is this the first of many patches that are needed, or is it the only one that you can foresee? I'd like to make sure that the maintenance overhead will be worth it going forwards. 

@cjappl have you got any concerns with introducing FreeBSD support?

devnexen commented:
I do not plan other commits, except maybe bug fixes. rtsan would have about the same potential usefulness than in macos.

cjappl commented:
> @devnexen in principle, I'm of course not opposed to rtsan supporting FreeBSD - is there any motivation in particular for why you propose adding it here? Is this the first of many patches that are needed, or is it the only one that you can foresee? I'd like to make sure that the maintenance overhead will be worth it going forwards.
> 
> @cjappl have you got any concerns with introducing FreeBSD support?

I have no specific concerns, but I agree we should be sure any maintenance cost is worth enabling it. Maybe we can keep this PR open and see if it attracts any comments by people wanting it?

I don't think we should enable it just to enable it. We have had strong signals of support on MacOS, Linux and (not implemented) Windows. We have not had any feedback asking for support on the BSD flavors yet.

brooksdavis commented:
rtsan seems like something we'd want on FreeBSD. If this is all that's required it seems like a no-brainer.

devnexen commented:
> > @devnexen in principle, I'm of course not opposed to rtsan supporting FreeBSD - is there any motivation in particular for why you propose adding it here? Is this the first of many patches that are needed, or is it the only one that you can foresee? I'd like to make sure that the maintenance overhead will be worth it going forwards.
> > @cjappl have you got any concerns with introducing FreeBSD support?
> 
> I have no specific concerns, but I agree we should be sure any maintenance cost is worth enabling it. Maybe we can keep this PR open and see if it attracts any comments by people wanting it?
> 
> I don't think we should enable it just to enable it. We have had strong signals of support on MacOS, Linux and (not implemented) Windows. We have not had any feedback asking for support on the BSD flavors yet.

Sure we can keep it open for a while, still a long road before 21.x anyway.

bsdimp commented:
> I don't think we should enable it just to enable it. We have had strong signals of support on MacOS, Linux and (not implemented) Windows. We have not had any feedback asking for support on the BSD flavors yet.

rtsan would be quite useful on BSD, especially FreeBSD. It's a highly desirable feature that we'd love to have.

cjappl reviewed (commented):
(sorry meant to leave these comments the other day, these comments are separate from the "should we enable it" discussion)

cjappl commented:
Thanks for the feedback @brooksdavis and @bsdimp - always hard to tell how popular or unpopular a change would be, especially from "outside the community" (aka me not knowing much about the BSD flavors :) 

> Sure we can keep it open for a while, still a long road before 21.x anyway.

Sounds like a plan.

allanjude commented:
We have seen interest in all of the different sanitizers from our embedded customers who use FreeBSD to build appliances. We have done work to extend support in FreeBSD for LLVM's KASAN and KMSAN, and have also seen commercial interest in KCSAN and RTSAN

devnexen commented on compiler-rt/lib/rtsan/rtsan_context.cpp:
```diff
@@ -43,6 +44,21 @@ static __rtsan::Context &GetContextForThisThreadImpl() {
+#else
+
+// On FreeBSD, pthread api cannot be used as calloc is called under the hood
+// at library initialization time.
```
I ll try to see what I can do later this week. basically even just compiling a binary doing nothing I get from lldb a recursive call on libpthread initialisation -> calloc, rinse and repeat ... then eventually crash.

devnexen replied:
So I know a bit more now. What I really need to do is either disabling mmap or using internal allocator before hands then it works, test pass in similar fashion as macos.

devnexen commented on compiler-rt/lib/rtsan/rtsan_context.cpp:
```diff
@@ -43,6 +44,21 @@ static __rtsan::Context &GetContextForThisThreadImpl() {
+static __rtsan::Context &GetContextForThisThreadImpl() {
+  if (ctx == nullptr) {
+    ctx = static_cast<Context *>(MmapOrDie(sizeof(Context), "RtsanContext"));
+    new (ctx) Context();
```
Yes I tried that originally but it freezes I ll try again or trying to make pthread api works.

cjappl replied:
My suspicion is that one of the interceptors is getting called by pthread, which tries to set up the context, which calls pthread, which calls the interceptor, which tries to set up the context............

I recommend figuring out which of those interceptors this is and checking if we can do an early out like this dlsym init early out:

https://github.com/llvm/llvm-project/blob/e223485c9b38a5579991b8cebb6a200153eee245/compiler-rt/lib/rtsan/rtsan_interceptors_posix.cpp#L776-L782

devnexen committed:
using pthread api for context creation.

but disabling mmap interception for this platform, which is the root
cause of the previous infinite call recursion.

cjappl commented on compiler-rt/lib/rtsan/rtsan_interceptors_posix.cpp:
```diff
@@ -864,11 +864,18 @@ INTERCEPTOR(void *, pvalloc, size_t size) {
 
+#if !SANITIZER_FREEBSD
+// enabling this interception on freebsd leads to infinite recursion
+// on pthread lib initialization
```
Can you do something like use internal_mmap here if we are initializing? If you check out rtsan.cpp there are three initializing states. 

May save us from disabling this on freebsd.

devnexen replied:
Unfortunately when mmap/calloc are hit rtsan seems already initialized, in that case internal_mmap is never called.

davidtrevelyan replied:
@devnexen would it be possible for you to share a stack trace so we can understand the route to infinite recursion during pthread init?

devnexen replied:
![image](https://github.com/user-attachments/assets/3665d61a-ad9c-4d7d-b617-aac6b639c4c7)

devnexen replied:
oh yes that s pthread_once but I ll give you the screenshot soon-ish.

davidtrevelyan replied:
@devnexen sorry - I meant the other end of the stack, not the point at which the segfault happens. Could you set a breakpoint in `pthread_once`, maybe, and share the backtrace again, so we can see the full initiating function call stack?

This is what I mean:

```lldb
⏺ build ❯ lldb ./projects/compiler-rt/lib/rtsan/tests/Rtsan-arm64-Test
(lldb) target create "./projects/compiler-rt/lib/rtsan/tests/Rtsan-arm64-Test"
Current executable set to '/Users/david/Repositories/rtsan/llvm-project/build/projects/compiler-rt/lib/rtsan/tests/Rtsan-arm64-Test' (arm64).
(lldb) br se -n pthread_once
Breakpoint 1: where = libsystem_pthread.dylib`pthread_once, address = 0x000000018044aff4
(lldb) pr la
Process 67543 launched: '/Users/david/Repositories/rtsan/llvm-project/build/projects/compiler-rt/lib/rtsan/tests/Rtsan-arm64-Test' (arm64)
Process 67543 stopped
* thread #1, stop reason = breakpoint 1.1
    frame #0: 0x000000018d10eff4 libsystem_pthread.dylib`pthread_once
libsystem_pthread.dylib`pthread_once:
->  0x18d10eff4 <+0>:  pacibsp
    0x18d10eff8 <+4>:  sub    sp, sp, #0x40
    0x18d10effc <+8>:  stp    x22, x21, [sp, #0x10]
    0x18d10f000 <+12>: stp    x20, x19, [sp, #0x20]
Target 0: (Rtsan-arm64-Test) stopped.
(lldb) bt
* thread #1, stop reason = breakpoint 1.1
  * frame #0: 0x000000018d10eff4 libsystem_pthread.dylib`pthread_once
    frame #1: 0x00000001002958b4 libclang_rt.rtsan_osx_dynamic.dylib`__rtsan::GetContextForThisThread() [inlined] GetContextForThisThreadImpl() at rtsan_context.cpp:34:3 [opt]
    frame #2: 0x00000001002958a0 libclang_rt.rtsan_osx_dynamic.dylib`__rtsan::GetContextForThisThread() at rtsan_context.cpp:62:10 [opt]
    frame #3: 0x000000010029545c libclang_rt.rtsan_osx_dynamic.dylib`__rtsan_notify_intercepted_call(func_name="shm_open") at rtsan.cpp:140:21 [opt]
    frame #4: 0x0000000100297364 libclang_rt.rtsan_osx_dynamic.dylib`wrap_shm_open(name="com.apple.featureflags.shm", oflag=0, mode=16832) at rtsan_interceptors_posix.cpp:973:3 [opt]
    frame #5: 0x000000018cfa41e8 libsystem_featureflags.dylib`_os_feature_table_once + 40
    frame #6: 0x000000018d1455e0 libsystem_platform.dylib`_os_once_callout + 32
    frame #7: 0x000000018cfa41bc libsystem_featureflags.dylib`_os_feature_table + 80
    frame #8: 0x000000018cfa51cc libsystem_featureflags.dylib`_os_feature_enabled_simple_impl + 44
    frame #9: 0x000000018cf29a10 libsystem_malloc.dylib`__malloc_init + 104
    frame #10: 0x000000019abd2604 libSystem.B.dylib`libSystem_initializer + 204
    frame #11: 0x000000018cda7d54 dyld`invocation function for block in dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const + 276
    frame #12: 0x000000018cde64d0 dyld`invocation function for block in dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 340
    frame #13: 0x000000018cdd9c38 dyld`invocation function for block in dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const + 496
    frame #14: 0x000000018cd8c2dc dyld`dyld3::MachOFile::forEachLoadCommand(Diagnostics&, void (load_command const*, bool&) block_pointer) const + 300
    frame #15: 0x000000018cdd8bcc dyld`dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const + 192
    frame #16: 0x000000018cde5fe4 dyld`dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 516
    frame #17: 0x000000018cda7bb4 dyld`dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const + 176
    frame #18: 0x000000018cdb2be8 dyld`dyld4::PrebuiltLoader::runInitializers(dyld4::RuntimeState&) const + 44
    frame #19: 0x000000018cdc78b8 dyld`dyld4::APIs::runAllInitializersForMain() + 84
    frame #20: 0x000000018cd918c8 dyld`dyld4::prepare(dyld4::APIs&, dyld3::MachOAnalyzer const*) + 3268
    frame #21: 0x000000018cd90bc0 dyld`dyld4::start(dyld4::KernelArgs*, void*, void*)::$_0::operator()() const + 544
    frame #22: 0x000000018cd9005c dyld`start + 2304
```

devnexen replied:
![image](https://github.com/user-attachments/assets/041d0bcc-8679-434f-bcb7-82ba83687c58)

davidtrevelyan replied:
I see - thanks. Does the problem go away if you put the `pthread_once` invocation into a disabled scope?

```cpp
static __rtsan::Context &GetContextForThisThreadImpl() {
  // ...
  {
    __rtsan::ScopedDisabler d;
    pthread_once(&key_once, MakThreadLocalContextKey);
  }
  // ...
```

I don't think this would be too unreasonable if the FreeBSD implementation of `pthread_once` allocates. I guess we can study the FreeBSD [source](https://github.com/freebsd/freebsd-src/blob/f8f5b658bc76cd539951fb8ac712d9acf71763ca/lib/libthr/thread/thr_once.c#L35) to find more details.

What do you think @cjappl? Disabling inside `GetContextForThisThread` make me a little uneasy as I haven't been able to consider all the implications. But I think it's potentially worth considering if it means we get to keep all interceptors on FreeBSD...

devnexen replied:
sure let me give a try.

cjappl replied:
It looks like dfsan does an early out if it isn't initialized, we could likely do the same?

https://github.com/llvm/llvm-project/blob/f451d27b387cdff14f0f45f1b3314090a5008e0c/compiler-rt/lib/dfsan/dfsan_interceptors.cpp#L132-L133


> What do you think @cjappl? Disabling inside GetContextForThisThread make me a little uneasy as I haven't been able to consider all the implications. But I think it's potentially worth considering if it means we get to keep all interceptors on FreeBSD...

I think I would rather (temporarily or permanently) lose mmap in freebsd than do the scoped disabler right now. I am worried about some side effect and losing an interceptor on one platform doesn't seem like the end of the world.

Let me ponder it a bit, curious if we could early out in mmap like dfsan

cjappl commented on compiler-rt/lib/rtsan/tests/rtsan_test_interceptors_posix.cpp:
```diff
@@ -1539,6 +1543,7 @@ TEST_F(KqueueTest, KeventDiesWhenRealtime) {
   ExpectNonRealtimeSurvival(Func);
 }
 
+#if SANITIZER_APPLE
```
"Kqueue is a scalable event notification interface introduced in FreeBSD 4.1 in July 2000, also supported in NetBSD, OpenBSD, DragonFly BSD, and macOS."

Perhaps we just want `!FREE_BSD` or similar. Also I would potentially un-nest these `ifdefs` for clarity

devnexen replied:
note that s the kevent64 flavour only.

cjappl replied:
ah, fine to leave then probably. Thanks for pointing that out

cjappl reviewed (commented):
Cool, much better with this version of the init! A few more smaller scale things and this is looking good

cjappl commented:
> Hi @cjappl did you have time to give any though since ? :)

Hi @devnexen,

@davidtrevelyan and I are working on a way to try to fix the reason that mmap needs to be disabled, as we think it is causing other problems elsewhere. So we appreciate your patience while we work on that, we will keep you posted.

I just started a job at a company that uses FreeBSD, so something like this will be valuable there before too long.
